bot/utils/redis_cache.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import annotations
+
+from typing import Any, AsyncIterator, Dict, Optional, Union
+
+from bot.bot import Bot
+
+TYPESTRING_PREFIXES = (
+    ("f|", float),
+    ("i|", int),
+    ("s|", str),
+)
+ValidRedisType = Union[str, int, float]
+
+
+class RedisCache:
+    """
+    A simplified interface for a Redis connection.
+
+    This class must be created as a class attribute in a class. This is because it
+    uses __set_name__ to create a namespace like MyCog.my_class_attribute which is
+    used as a hash name when we store stuff in Redis, to prevent collisions.
+
+    The class this object is instantiated in must also contains an attribute with an
+    instance of Bot. This is because Bot contains our redis_pool, which is how this
+    class communicates with the Redis server.
```

Comment:
On the fence here, but these paragraphs should be less specific. I don't think they need to mention implementation details like the fact that it uses `__set_name__` or that it relies on the Bot specifically for the redis pool. If you still want this information somewhere, then it's better to put it in some internal docs rather than the class's docstring.

Reply:
I don't think I agree with that. Docstrings _are_ internal docs. I think I have the opposite opinion here: I don't particularly see the need for some guide on our website to contain these sorts of implementation details, a guide should just explain how to use it. Docstrings, however, are for people who are digging into the source, and should provide a fuller idea of how the object works and behaves.

I should fix the wrong variable name, though.

Reply:
Okay, as a compromise I've moved some of the specifics into the `__set_name__` and `__get__` docstrings, and simplified the stuff in the class docstring to this:

```
Please note that this class MUST be created as a class attribute, and that that class
must also contain an attribute with an instance of our Bot. See `__get__` and `__set_name__`
for more information about how this works.
```

Reply:
If I want to learn how to use this class, I will first read the class docstring to get an overview. My perspective is that knowing the implementation details is just noise at this point. I don't need to know that it specifically uses `__set_name__` in order to learn to effectively use this class; I just need to know that it has to be instantiated as a class attribute.

I don't know what you mean by website guide. By internal docs, I meant moving this information to one of the "private" methods in the class, like to `__set_name__` itself or `_set_namespace`.

Reply:
Yeah - that's what I ended up doing, so I guess this is good then?